File: shell_wrapper/ahe_test.cc
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "shell_wrapper/ahe.h"

#include <cstddef>
#include <cstdint>
#include <iterator>
#include <memory>
#include <string>
#include <utility>
#include <vector>

#include "absl/strings/string_view.h"
#include "gmock/gmock.h"
#include "gtest/gtest.h"
#include "shell_encryption/rns/testing/testing_utils.h"
#include "shell_wrapper/ahe.rs.h"
#include "shell_wrapper/shell_aliases.h"
#include "shell_wrapper/shell_types.h"
#include "shell_wrapper/shell_types.rs.h"
#include "shell_wrapper/single_thread_hkdf.h"
#include "shell_wrapper/single_thread_hkdf.rs.h"
#include "shell_wrapper/status.h"
#include "shell_wrapper/status_matchers.h"

namespace secure_aggregation {
namespace {

using ::testing::HasSubstr;

constexpr int kLogN = 11;
constexpr int kT = 10001;
constexpr uint64_t kQs[] = {1073692673, 1073668097};
constexpr uint64_t kQsLarge[] = {1125899906826241, 1125899906629633};
constexpr double kSFlood = 1.0e+10;

constexpr int kNumParties = 3;
constexpr int kPublicKeyVariance = 8;
constexpr uint64_t kMaxValue = 72;
constexpr double kSBase = 12.8;

TEST(AheTest, EncryptDecryptOne) {
  // Create the public parameters.
  std::unique_ptr<std::string> public_seed;
  auto status = GenerateSingleThreadHkdfSeed(public_seed);
  SECAGG_ASSERT_OK(UnwrapFfiStatus(status));
  AhePublicParameters public_params;
  SECAGG_ASSERT_OK(UnwrapFfiStatus(CreateAhePublicParameters(
      kLogN, kT, kQs, std::size(kQs), kPublicKeyVariance, kSBase, kSFlood,
      ToRustSlice(*public_seed), &public_params)));

  std::unique_ptr<std::string> private_seed;
  status = GenerateSingleThreadHkdfSeed(private_seed);
  SECAGG_ASSERT_OK(UnwrapFfiStatus(status));
  SingleThreadHkdfWrapper prng;
  status = CreateSingleThreadHkdf(ToRustSlice(*private_seed), prng);
  SECAGG_ASSERT_OK(UnwrapFfiStatus(status));

  RnsPolynomialWrapper sk_share;
  SECAGG_ASSERT_OK(
      UnwrapFfiStatus(GenerateSecretKeyShare(public_params, &prng, &sk_share)));

  RnsPolynomialWrapper public_key_b = CreateEmptyRnsPolynomialWrapper();
  RnsPolynomialWrapper public_key_share_error =
      CreateEmptyRnsPolynomialWrapper();
  SECAGG_ASSERT_OK(UnwrapFfiStatus(GeneratePublicKeyShareWrapper(
      sk_share, public_params, &prng, &public_key_b, &public_key_share_error,
      nullptr)));

  // Encrypt one vector.
  const int num_messages = 1 << kLogN;
  std::vector<Integer> coeffs0 =
      rlwe::testing::SampleMessages(num_messages, kMaxValue);

  auto ciphertext_b = CreateEmptyRnsPolynomialWrapper();
  auto ciphertext_a = CreateEmptyRnsPolynomialWrapper();
  auto c0_r = CreateEmptyRnsPolynomialWrapper();
  auto c0_e = CreateEmptyRnsPolynomialWrapper();

  SECAGG_ASSERT_OK(UnwrapFfiStatus(AheEncrypt(
      coeffs0.data(), coeffs0.size(), public_key_b, public_params, &prng,
      &ciphertext_b, &ciphertext_a, &c0_r, &c0_e, /*wraparound=*/nullptr)));

  RnsPolynomialWrapper partial_decryption;
  SECAGG_ASSERT_OK(UnwrapFfiStatus(
      PartialDecrypt(ciphertext_a, sk_share, public_params, &prng,
                     &partial_decryption, /*error_flood=*/nullptr,
                     /*wraparound=*/nullptr)));

  // Recover from partial decryptions and component b
  uint64_t decrypted_buffer[2 * num_messages];
  size_t n_written;
  SECAGG_ASSERT_OK(UnwrapFfiStatus(RecoverMessages(
      partial_decryption, ciphertext_b, public_params,
      std::size(decrypted_buffer), decrypted_buffer, &n_written)));

  for (int i = 0; i < 2; ++i) {
    Integer decrypted_coeff = decrypted_buffer[i];
    Integer expected = coeffs0[i];
    EXPECT_EQ(decrypted_coeff, expected);
  }
}

// Extern C version of
// https://github.com/google/shell-encryption/blob/master/shell_encryption/multi_party/recovery_test.cc
TEST(AheTest, ExternCRecoveryTest) {
  // Create the public parameters.
  std::unique_ptr<std::string> public_seed;
  auto status = GenerateSingleThreadHkdfSeed(public_seed);
  SECAGG_ASSERT_OK(UnwrapFfiStatus(status));
  AhePublicParameters public_params;
  SECAGG_ASSERT_OK(UnwrapFfiStatus(CreateAhePublicParameters(
      kLogN, kT, kQs, std::size(kQs), kPublicKeyVariance, kSBase, kSFlood,
      ToRustSlice(*public_seed), &public_params)));

  std::unique_ptr<std::string> private_seed;
  status = GenerateSingleThreadHkdfSeed(private_seed);
  SECAGG_ASSERT_OK(UnwrapFfiStatus(status));
  SingleThreadHkdfWrapper prng;
  status = CreateSingleThreadHkdf(ToRustSlice(*private_seed), prng);
  SECAGG_ASSERT_OK(UnwrapFfiStatus(status));

  std::vector<RnsPolynomialWrapper> secret_key_shares;
  std::vector<RnsPolynomialWrapper> public_key_shares;
  secret_key_shares.reserve(kNumParties);
  public_key_shares.reserve(kNumParties);
  for (int i = 0; i < kNumParties; ++i) {
    RnsPolynomialWrapper sk_share;
    SECAGG_ASSERT_OK(UnwrapFfiStatus(
        GenerateSecretKeyShare(public_params, &prng, &sk_share)));

    RnsPolynomialWrapper public_key_share_b = CreateEmptyRnsPolynomialWrapper();
    RnsPolynomialWrapper public_key_share_error =
        CreateEmptyRnsPolynomialWrapper();
    RnsPolynomialWrapper wraparound = CreateEmptyRnsPolynomialWrapper();
    SECAGG_ASSERT_OK(UnwrapFfiStatus(GeneratePublicKeyShareWrapper(
        sk_share, public_params, &prng, &public_key_share_b,
        &public_key_share_error, &wraparound)));
    secret_key_shares.push_back(std::move(sk_share));
    public_key_shares.push_back(std::move(public_key_share_b));
  }

  // Create public key by aggregating public key shares.
  RnsPolynomialWrapper public_key_b;
  SECAGG_ASSERT_OK(UnwrapFfiStatus(
      CreateZeroRnsPolynomialWrapper(public_params, &public_key_b)));
  for (const auto& public_key_share : public_key_shares) {
    SECAGG_ASSERT_OK(UnwrapFfiStatus(
        AddInPlace(public_params, public_key_share, &public_key_b)));
  }

  // Encrypt two vectors.
  const int num_messages = 1 << kLogN;
  std::vector<Integer> coeffs0 =
      rlwe::testing::SampleMessages(num_messages, kMaxValue);
  std::vector<Integer> coeffs1 =
      rlwe::testing::SampleMessages(num_messages, kMaxValue);

  auto c0_b = CreateEmptyRnsPolynomialWrapper();
  auto c0_a = CreateEmptyRnsPolynomialWrapper();
  auto c0_r = CreateEmptyRnsPolynomialWrapper();
  auto c0_e = CreateEmptyRnsPolynomialWrapper();
  SECAGG_ASSERT_OK(UnwrapFfiStatus(AheEncrypt(coeffs0.data(), coeffs0.size(),
                                              public_key_b, public_params,
                                              &prng, &c0_b, &c0_a, &c0_r, &c0_e,
                                              /*wraparound=*/nullptr)));

  auto c1_b = CreateEmptyRnsPolynomialWrapper();
  auto c1_a = CreateEmptyRnsPolynomialWrapper();
  auto c1_r = CreateEmptyRnsPolynomialWrapper();
  auto c1_e = CreateEmptyRnsPolynomialWrapper();
  SECAGG_ASSERT_OK(UnwrapFfiStatus(AheEncrypt(coeffs1.data(), coeffs1.size(),
                                              public_key_b, public_params,
                                              &prng, &c1_b, &c1_a, &c1_r, &c1_e,
                                              /*wraparound=*/nullptr)));

  // Accumulate component a
  RnsPolynomialWrapper ciphertext_a;
  SECAGG_ASSERT_OK(UnwrapFfiStatus(
      CreateZeroRnsPolynomialWrapper(public_params, &ciphertext_a)));
  SECAGG_ASSERT_OK(
      UnwrapFfiStatus(AddInPlace(public_params, c0_a, &ciphertext_a)));
  SECAGG_ASSERT_OK(
      UnwrapFfiStatus(AddInPlace(public_params, c1_a, &ciphertext_a)));

  // Let all secret key share holders do partial decryption.
  std::vector<RnsPolynomialWrapper> partial_decryptions;
  partial_decryptions.reserve(kNumParties);
  for (int i = 0; i < kNumParties; ++i) {
    RnsPolynomialWrapper partial_decryption;
    SECAGG_ASSERT_OK(UnwrapFfiStatus(
        PartialDecrypt(ciphertext_a, secret_key_shares[i], public_params, &prng,
                       &partial_decryption,
                       /*error_flood=*/nullptr,
                       /*wraparound=*/nullptr)));
    partial_decryptions.push_back(std::move(partial_decryption));
  }

  // Accumulate partial decryptions
  RnsPolynomialWrapper sum_partial_decryptions;
  SECAGG_ASSERT_OK(UnwrapFfiStatus(
      CreateZeroRnsPolynomialWrapper(public_params, &sum_partial_decryptions)));
  for (const auto& partial_decryption : partial_decryptions) {
    SECAGG_ASSERT_OK(UnwrapFfiStatus(AddInPlace(
        public_params, partial_decryption, &sum_partial_decryptions)));
  }

  // Accumulate component b
  RnsPolynomialWrapper ciphertext_b;
  SECAGG_ASSERT_OK(UnwrapFfiStatus(
      CreateZeroRnsPolynomialWrapper(public_params, &ciphertext_b)));
  SECAGG_ASSERT_OK(
      UnwrapFfiStatus(AddInPlace(public_params, c0_b, &ciphertext_b)));
  SECAGG_ASSERT_OK(
      UnwrapFfiStatus(AddInPlace(public_params, c1_b, &ciphertext_b)));

  // Recover from partial decryptions and component b
  uint64_t decrypted_buffer[2 * num_messages];
  size_t n_written;
  SECAGG_ASSERT_OK(UnwrapFfiStatus(RecoverMessages(
      sum_partial_decryptions, ciphertext_b, public_params,
      std::size(decrypted_buffer), decrypted_buffer, &n_written)));
  EXPECT_EQ(n_written, num_messages);

  EXPECT_EQ(public_params.encoder->PlaintextModulus(), kT);

  for (int i = 0; i < num_messages; ++i) {
    Integer decrypted_coeff = decrypted_buffer[i];
    Integer expected = (coeffs0[i] + coeffs1[i]) % kT;
    EXPECT_EQ(decrypted_coeff, expected);
  }
}

}  // namespace
}  // namespace secure_aggregation

File: shell_wrapper/ahe_test.rs
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use ahe::{
    ahe_encrypt, create_public_parameters, create_zero_rns_polynomial, generate_public_key_share,
    generate_secret_key, get_moduli, get_rns_context_ref, partial_decrypt, public_key_component_a,
    recover_messages,
};
use googletest::{
    expect_that, gtest,
    matchers::{contains, eq, not},
    Result,
};
use rand::Rng;
use shell_types::{add_in_place, create_empty_rns_polynomial, write_rns_polynomial_to_buffer_128};

const LOG_N: u64 = 11;
const T: u64 = 10001;
const QS: [u64; 2] = [1073692673, 1073668097];
const QS_LARGE: [u64; 2] = [1125899906826241, 1125899906629633];
const S_FLOOD: f64 = 1.0e+10;

const NUM_PARTIES: usize = 5;
const PUBLIC_KEY_VARIANCE: u64 = 8;
const MAX_VALUE: u64 = 72;
const S_BASE: f64 = 12.8;

#[gtest]
fn encrypt_add_recover() -> Result<()> {
    // Create the public parameters.
    let public_seed = single_thread_hkdf::generate_seed()?;
    let params = create_public_parameters(
        LOG_N,
        T,
        &QS,
        PUBLIC_KEY_VARIANCE,
        S_BASE,
        S_FLOOD,
        &public_seed,
    )?;
    let moduli = get_moduli(&params);

    let private_seed = single_thread_hkdf::generate_seed()?;
    let mut prng = single_thread_hkdf::create(&private_seed)?;

    // Distributed key generation.
    let mut secret_key_shares = Vec::new();
    let mut public_key_shares = Vec::new();
    for _ in 0..NUM_PARTIES {
        let sk_share = generate_secret_key(&params, &mut prng)?;

        let mut public_key_share_b = create_empty_rns_polynomial();
        let mut public_key_share_error = create_empty_rns_polynomial();
        generate_public_key_share(
            &sk_share,
            &params,
            &mut prng,
            &mut public_key_share_b,
            &mut public_key_share_error,
            None,
        )?;

        secret_key_shares.push(sk_share);
        public_key_shares.push(public_key_share_b);
    }
    let mut public_key_b = create_zero_rns_polynomial(&params)?;
    for public_key_share in public_key_shares {
        add_in_place(&moduli, &public_key_share, &mut public_key_b)?;
    }

    // Generate two random vectors.
    let num_messages = 1 << LOG_N;
    let mut coeffs0: Vec<u64> = Vec::with_capacity(num_messages);
    let mut coeffs1: Vec<u64> = Vec::with_capacity(num_messages);
    for _ in 0..num_messages {
        coeffs0.push(rand::thread_rng().gen_range(0..MAX_VALUE));
        coeffs1.push(rand::thread_rng().gen_range(0..MAX_VALUE));
    }

    // Encrypt the two vectors.
    let mut c0_b = create_empty_rns_polynomial();
    let mut c0_a = create_empty_rns_polynomial();
    let mut c0_r = create_empty_rns_polynomial();
    let mut c0_e = create_empty_rns_polynomial();
    ahe_encrypt(
        &coeffs0,
        &public_key_b,
        &params,
        &mut prng,
        &mut c0_b,
        &mut c0_a,
        &mut c0_r,
        &mut c0_e,
        None,
    )?;

    let mut c1_b = create_empty_rns_polynomial();
    let mut c1_a = create_empty_rns_polynomial();
    let mut c1_r = create_empty_rns_polynomial();
    let mut c1_e = create_empty_rns_polynomial();
    ahe_encrypt(
        &coeffs1,
        &public_key_b,
        &params,
        &mut prng,
        &mut c1_b,
        &mut c1_a,
        &mut c1_r,
        &mut c1_e,
        None,
    )?;

    // Accumulate component a
    let mut ciphertext_a = create_zero_rns_polynomial(&params)?;
    add_in_place(&moduli, &c0_a, &mut ciphertext_a)?;
    add_in_place(&moduli, &c1_a, &mut ciphertext_a)?;

    // Let all secret key share holders do partial decryption.
    let mut sum_partial_decryptions = create_zero_rns_polynomial(&params)?;
    for sk_share in secret_key_shares {
        let partial_decryption =
            partial_decrypt(&ciphertext_a, &sk_share, &params, &mut prng, None, None)?;
        add_in_place(&moduli, &partial_decryption, &mut sum_partial_decryptions)?;
    }

    // Accumulate component b
    let mut ciphertext_b = create_zero_rns_polynomial(&params)?;
    add_in_place(&moduli, &c0_b, &mut ciphertext_b)?;
    add_in_place(&moduli, &c1_b, &mut ciphertext_b)?;

    // Recover from partial decryptions and component b
    let mut output_values = vec![0; 2 * num_messages];
    let n_written =
        recover_messages(&sum_partial_decryptions, &ciphertext_b, &params, &mut output_values)?;
    expect_that!(n_written, eq(num_messages));

    // Check homomorphism.
    for i in 0..num_messages {
        expect_that!(output_values[i], eq((coeffs0[i] + coeffs1[i]) % T));
    }

    Ok(())
}

#[gtest]
fn export_public_key_component_a_not_trivially_broken() -> Result<()> {
    // Create the public parameters.
    let public_seed = single_thread_hkdf::generate_seed()?;
    let params = create_public_parameters(
        LOG_N,
        T,
        &QS_LARGE,
        PUBLIC_KEY_VARIANCE,
        S_BASE,
        S_FLOOD,
        &public_seed,
    )?;
    let rns_context = get_rns_context_ref(&params);

    let pk_a = public_key_component_a(&params)?;
    let mut buf = vec![0u64; 1 << (LOG_N + 1)];
    write_rns_polynomial_to_buffer_128(&rns_context, &pk_a, &mut buf)?;

    // Check that none of the public key coefficients are zero or less than 64 bits.
    expect_that!(buf, not(contains(eq(&0u64))));
    Ok(())
}

#[gtest]
fn generate_public_key_share_writes_to_wraparound() -> Result<()> {
    // Create the public parameters.
    let public_seed = single_thread_hkdf::generate_seed()?;
    let params = create_public_parameters(
        LOG_N,
        T,
        &QS_LARGE,
        PUBLIC_KEY_VARIANCE,
        S_BASE,
        S_FLOOD,
        &public_seed,
    )?;
    let private_seed = single_thread_hkdf::generate_seed()?;
    let mut prng = single_thread_hkdf::create(&private_seed)?;

    let sk_share = generate_secret_key(&params, &mut prng)?;

    let mut public_key_share_b = create_empty_rns_polynomial();
    let mut public_key_share_error = create_empty_rns_polynomial();
    let mut wraparound = create_empty_rns_polynomial();
    generate_public_key_share(
        &sk_share,
        &params,
        &mut prng,
        &mut public_key_share_b,
        &mut public_key_share_error,
        Some(&mut wraparound),
    )?;

    let rns_context = get_rns_context_ref(&params);

    let mut buf = vec![0u64; 1 << (LOG_N + 1)];
    // Check that the wraparound is not empty.
    write_rns_polynomial_to_buffer_128(&rns_context, &wraparound, &mut buf)?;
    let mut wraparound_is_empty = true;
    for i in 0..1 << (LOG_N + 1) {
        if buf[i] != 0 {
            wraparound_is_empty = false;
            break;
        }
    }
    expect_that!(wraparound_is_empty, eq(false));
    Ok(())
}

#[gtest]
fn generate_ahe_encrypt_writes_to_wraparound() -> Result<()> {
    // Create the public parameters.
    let public_seed = single_thread_hkdf::generate_seed()?;
    let params = create_public_parameters(
        LOG_N,
        T,
        &QS_LARGE,
        PUBLIC_KEY_VARIANCE,
        S_BASE,
        S_FLOOD,
        &public_seed,
    )?;
    let private_seed = single_thread_hkdf::generate_seed()?;
    let mut prng = single_thread_hkdf::create(&private_seed)?;

    let sk_share = generate_secret_key(&params, &mut prng)?;

    let mut public_key_share_b = create_empty_rns_polynomial();
    let mut public_key_share_error = create_empty_rns_polynomial();
    generate_public_key_share(
        &sk_share,
        &params,
        &mut prng,
        &mut public_key_share_b,
        &mut public_key_share_error,
        None,
    )?;

    let mut coeffs = vec![0u64; 8];
    for _ in 0..8 {
        coeffs.push(rand::thread_rng().gen_range(0..MAX_VALUE));
    }
    let mut ct_b = create_empty_rns_polynomial();
    let mut ct_a = create_empty_rns_polynomial();
    let mut ct_r = create_empty_rns_polynomial();
    let mut ct_e = create_empty_rns_polynomial();
    let mut wraparound = create_empty_rns_polynomial();
    ahe_encrypt(
        &coeffs,
        &public_key_share_b,
        &params,
        &mut prng,
        &mut ct_b,
        &mut ct_a,
        &mut ct_r,
        &mut ct_e,
        Some(&mut wraparound),
    )?;

    let rns_context = get_rns_context_ref(&params);

    let mut buf = vec![0u64; 1 << (LOG_N + 1)];
    // Check that the wraparound is not empty.
    write_rns_polynomial_to_buffer_128(&rns_context, &wraparound, &mut buf)?;
    let mut wraparound_is_empty = true;
    for i in 0..1 << (LOG_N + 1) {
        if buf[i] != 0 {
            wraparound_is_empty = false;
            break;
        }
    }
    expect_that!(wraparound_is_empty, eq(false));
    Ok(())
}

#[gtest]
fn generate_partial_decrypt_writes_to_error_flood_and_wraparound() -> Result<()> {
    // Create the public parameters.
    let public_seed = single_thread_hkdf::generate_seed()?;
    let params = create_public_parameters(
        LOG_N,
        T,
        &QS_LARGE,
        PUBLIC_KEY_VARIANCE,
        S_BASE,
        S_FLOOD,
        &public_seed,
    )?;
    let private_seed = single_thread_hkdf::generate_seed()?;
    let mut prng = single_thread_hkdf::create(&private_seed)?;

    let sk_share = generate_secret_key(&params, &mut prng)?;

    let mut public_key_share_b = create_empty_rns_polynomial();
    let mut public_key_share_error = create_empty_rns_polynomial();
    generate_public_key_share(
        &sk_share,
        &params,
        &mut prng,
        &mut public_key_share_b,
        &mut public_key_share_error,
        None,
    )?;

    let mut coeffs = vec![0u64; 8];
    for _ in 0..8 {
        coeffs.push(rand::thread_rng().gen_range(0..MAX_VALUE));
    }
    let mut ct_b = create_empty_rns_polynomial();
    let mut ct_a = create_empty_rns_polynomial();
    let mut ct_r = create_empty_rns_polynomial();
    let mut ct_e = create_empty_rns_polynomial();
    ahe_encrypt(
        &coeffs,
        &public_key_share_b,
        &params,
        &mut prng,
        &mut ct_b,
        &mut ct_a,
        &mut ct_r,
        &mut ct_e,
        None,
    )?;

    let mut error_flood = create_empty_rns_polynomial();
    let mut wraparound = create_empty_rns_polynomial();
    partial_decrypt(
        &ct_a,
        &sk_share,
        &params,
        &mut prng,
        Some(&mut error_flood),
        Some(&mut wraparound),
    )?;

    let rns_context = get_rns_context_ref(&params);

    let mut buf = vec![0u64; 1 << (LOG_N + 1)];
    // Check that the error flood is not empty.
    write_rns_polynomial_to_buffer_128(&rns_context, &error_flood, &mut buf)?;
    let mut error_flood_is_empty = true;
    for i in 0..1 << (LOG_N + 1) {
        if buf[i] != 0 {
            error_flood_is_empty = false;
            break;
        }
    }
    expect_that!(error_flood_is_empty, eq(false));

    // Check that the wraparound is not empty.
    write_rns_polynomial_to_buffer_128(&rns_context, &wraparound, &mut buf)?;
    let mut wraparound_is_empty = true;
    for i in 0..1 << (LOG_N + 1) {
        if buf[i] != 0 {
            wraparound_is_empty = false;
            break;
        }
    }
    expect_that!(wraparound_is_empty, eq(false));
    Ok(())
}
